README.md

```diff
@@ -34,25 +34,41 @@ which can be verified by computing the gradient at `(1.1, 2.2, 3.3)`:
 (2.4396285219055063, -1.953374825727421, 0.9654825811012627)
 ```
 
-When `foo` is instantiated to matrices, which is a similarly trivial example as before due to the arithmetic operations working on the arrays element-wise, the gradient is:
+We can instantiate `foo` to matrices; the operations within (`sin`, `+`, `*`) etc applying elementwise:
 ```hs
-type Matrix2x2 f r = f (TKS '[2, 2] r)
+type Matrix2x2 f r = f (TKS '[2, 2] r) -- AWF: TKS means [...?]
 type ThreeMatrices r = (Matrix2x2 Concrete r, Matrix2x2 Concrete r, Matrix2x2 Concrete r)
 threeSimpleMatrices :: ThreeMatrices Double
-threeSimpleMatrices = (srepl 1.1, srepl 2.2, srepl 3.3)
-gradFooMatrix :: (Differentiable r, GoodScalar r)
+threeSimpleMatrices = (srepl 1.1, srepl 2.2, srepl 3.3) -- srepl duplicates its argument in each matrix entry
+foo_value :: Matrix2x2 Concrete Double = foo threeSimpleMatrices
+```
+We can compute a gradient of a function from matrices to scalar, say `sumfoo`:
+```hs
+sumfoo :: ThreeMatrices r -> r
+sumfoo = kfromS . ssum0 . foo
+>>> sumfoo threeSimpleMatrices
+1.23456789 ????
+```
+and its gradient is given by
+```hs
+gradsumfoo :: (Differentiable r, GoodScalar r)
               => ThreeMatrices r -> ThreeMatrices r
-gradFooMatrix = cgrad (kfromS . ssum0 . foo)
+gradsumfoo = cgrad sumfoo
 ```
 
-where we had to augment function `foo`, because `cgrad` expects a function with a scalar codomain (e.g., a loss function for neural networks). This works as well as before:
+This works as well as before:
 ```hs
->>> gradFooMatrix threeSimpleMatrices
+>>> gradsumfoo threeSimpleMatrices
 (sfromListLinear [2,2] [2.4396285219055063,2.4396285219055063,2.4396285219055063,2.4396285219055063],sfromListLinear [2,2] [-1.953374825727421,-1.953374825727421,-1.953374825727421,-1.953374825727421],sfromListLinear [2,2] [0.9654825811012627,0.9654825811012627,0.9654825811012627,0.9654825811012627])
 ```
 
-Note that `w` is processed only once during gradient computation and this property of sharing preservation is guaranteed for the `cgrad` tool universally, without any action required from the user. When computing symbolic derivative programs, however, the user has to explicitly mark values for sharing using `tlet` with a more specific type of the objective function, as shown below.
+[AWF: better explain `GoodScalar` name somewhere here]
+
+### Efficiency: preserving sharing
 
+We noted above that `w` appears twice in `foo`.  A property of tracing-based AD systems is that such re-use may not be captured, with explosive results.
+In `cgrad`, such sharing is preserved, so `w` is processed only once during gradient computation and this property is guaranteed for the `cgrad` tool universally, without any action required from the user.
+When computing symbolic derivative programs, however, the user has to explicitly mark values for sharing using `tlet` with a more specific type of the objective function, as shown below.
```

Comment:
```suggestion
`horde-ad` also allows computing _symbolic_ derivative programs: using this API, a program is differentiated only once, after which it can be run on many different input values.
In this case, however, sharing is _not_ automatically preserved, so shared variables have to be explicitly marked using `tlet`, as shown below in `fooLet`.
This also makes the type of the function more specific: it now does not work on an arbitrary `Num` any more, but instead on an arbitrary `horde-ad` tensor that implements arithmetic operations like `atan2H`.
```
How should we explain the difference between `atan2` and `atan2H`?

Reply:
I think your remark about `atan2H` is sufficient (and in reality it's not different from `atan2` --- it just belongs to a class that contains solely `atan2H` and doesn't have a lot of silly superclasses). This is getting quite long already, so I'd keep it as you wrote it and maybe I'd even try to trim the README later on and move some of the contents to a hypothetical future tutorial.